grid/src/grid.rs

```diff
@@ -1,4 +1,432 @@
+        let mut data = AdjacencyList::<usize>::new();
+        let cells = &self.connectivity[2][0];
+        let edges = &self.connectivity[1][0];
+        for cell_type in [
+            ReferenceCellType::Triangle,
+            ReferenceCellType::Quadrilateral,
+        ] {
+            let ref_cell = get_reference_cell(cell_type);
+            let ref_edges = (0..ref_cell.edge_count())
+                .map(|x| ref_cell.connectivity(1, x, 0).unwrap())
+                .collect::<Vec<Vec<usize>>>();
+            for i in self.get_cells_range(cell_type).unwrap() {
+                let cell = unsafe { cells.row_unchecked(i) };
+                let mut row = vec![];
+                for e in &ref_edges {
+                    let start = min(cell[e[0]], cell[e[1]]);
+                    let end = max(cell[e[0]], cell[e[1]]);
+                    for (i, edge) in edges.iter_rows().enumerate() {
+                        if edge[0] == start && edge[1] == end {
+                            row.push(i);
+                            break;
+                        }
+                    }
+                }
+                data.add_row(&row);
+            }
+        }
+        self.connectivity[2][1] = data;
+    }
+    fn create_connectivity_22(&mut self) {
+        for i in 0..self.connectivity[2][0].num_rows() {
+            self.connectivity[2][2].add_row(&[i]);
+        }
+    }
+}
+
+impl Topology for Serial2DTopology {
+    fn index_map(&self) -> &[usize] {
+        &self.index_map
+    }
+    fn get_cells(&self, cell_type: ReferenceCellType) -> Vec<usize> {
+        match cell_type {
+            ReferenceCellType::Triangle => {
+                let mut cells = vec![];
+                for (i, cell) in self.connectivity[2][0].iter_rows().enumerate() {
+                    if cell.len() == 3 {
+                        cells.push(i);
+                    }
+                }
+                cells
+            }
+            ReferenceCellType::Quadrilateral => {
+                let mut cells = vec![];
+                for (i, cell) in self.connectivity[2][0].iter_rows().enumerate() {
+                    if cell.len() == 4 {
+                        cells.push(i);
+                    }
+                }
+                cells
+            }
+            _ => vec![],
+        }
+    }
+    fn get_cells_range(&self, cell_type: ReferenceCellType) -> Option<Range<usize>> {
+        match cell_type {
+            ReferenceCellType::Triangle => Some(0..self.quad_start),
+            ReferenceCellType::Quadrilateral => Some(self.quad_start..self.index_map.len()),
+            _ => Some(0..0),
+        }
+    }
+    fn local2global(&self, local_id: usize) -> usize {
+        self.index_map[local_id]
+    }
+    fn global2local(&self, global_id: usize) -> Option<usize> {
+        for (i, j) in self.index_map.iter().enumerate() {
+            if *j == global_id {
+                return Some(i);
+            }
+        }
+        None
+    }
```

Comment:
We have a similar function in PETSc (we keep them sorted, but it's still a binary search per index), but I think using it is an anti-pattern.